website/server/libs/groupTasks.js

```diff
@@ -1,61 +1,268 @@
+  );
+}
+
+async function groupTaskCompleted (groupMemberTask, user, now) {
+  const masterTask = await Tasks.Task.findOne({
+    _id: groupMemberTask.group.taskId,
+  }).exec();
+
+  if (!masterTask || !masterTask.group) return false;
+
+  if (masterTask.history && masterTask.history.length > 0) {
+    // I hope we don't really need to loop the history and check every date
+    // (User with a late day start in Australia vs and early one in Hawaii?)
+    const taskLastHistory = masterTask.history[masterTask.history.length - 1];
+    // Check if the history entry has a user who completed the task
+    // TODO When group scoring is implemented, add history entries for group missed dailies
+    if (taskLastHistory.userId) {
+      const lastCompletingUser = await Users.findById(taskLastHistory.userId, 'preferences');
+      if (lastCompletingUser) {
+        // Check what the completing user's "day" was when the task was completed
+        const taskLastCompletedDay = startOfDay(
+          defaults(
+            { now: moment(taskLastHistory.date) },
+            lastCompletingUser.preferences.toObject(),
+          ),
+        );
+        const userDay = startOfDay(defaults({ now }, user.preferences.toObject()));
+        if (userDay.isSame(taskLastCompletedDay)) return true;
+      }
+    }
   }
+
+  // Did not find a matching completion or valid completing user
+  return false;
 }
 
 async function handleSharedCompletion (groupMemberTask) {
   const masterTask = await Tasks.Task.findOne({
     _id: groupMemberTask.group.taskId,
   }).exec();
 
-  if (!masterTask || !masterTask.group || masterTask.type !== 'todo') return;
+  if (!masterTask || !masterTask.group || masterTask.type === 'habit') return;
 
   if (masterTask.group.sharedCompletion === SHARED_COMPLETION.single) {
-    await _deleteUnfinishedTasks(groupMemberTask);
-    await _completeMasterTask(masterTask);
+    await _updateAssignedUsersTasks(masterTask, groupMemberTask);
+    await _completeOrUncompleteMasterTask(
+      masterTask, groupMemberTask.completed
+      || (groupMemberTask.group.approval && groupMemberTask.group.approval.approved),
+    );
   } else if (masterTask.group.sharedCompletion === SHARED_COMPLETION.every) {
-    await _evaluateAllAssignedCompletion(masterTask);
+    await _evaluateAllAssignedCompletion(masterTask, groupMemberTask);
+  }
+}
+
+async function groupTaskNewDay (groupMemberTask, user, now = moment()) {
+  // Similar logic to groupTaskCompleted but returns update to uncomplete group task
+  // if user day is *different* from task's last completed day
+  const masterTask = await Tasks.Task.findOne({
+    _id: groupMemberTask.group.taskId,
+    completed: true,
+    type: 'daily',
+  }).exec();
+
+  if (!masterTask) return;
+
+  if (masterTask.history && masterTask.history.length > 0) {
+    const taskLastHistory = masterTask.history[masterTask.history.length - 1];
+    if (taskLastHistory.userId) {
```

Comment:
the logic here between `groupTaskNewDay ` and `groupTaskCompleted ` is very similar, maybe we can combine the two or share part of the code?